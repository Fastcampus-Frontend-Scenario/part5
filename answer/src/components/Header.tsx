import { Box, Drawer } from "@mui/material"
import MenuIcon from '@mui/icons-material/Menu';
import PersonOutlinedIcon from '@mui/icons-material/PersonOutlined'
import ShoppingCartOutlinedIcon from '@mui/icons-material/ShoppingCartOutlined';
import { useState } from "react";
import styled from "@emotion/styled";
import { useRouter } from "next/router";
export const Header: React.FC = () => {
    const [open, setOpen] = useState(false)
    const router = useRouter()
    const onMenuClick = () => {
        // TODO: show sidebar
    }
    return (
        <Container>
            <MenuIcon onClick={() => setOpen(true)} data-testid='menu-button' />
            <Box gap={'20px'} display={'flex'}>
                <ShoppingCartOutlinedIcon data-testid='cart-button' onClick={() => router.push('/cart')}/>
                <PersonOutlinedIcon data-testid='mypage-button' onClick={() => router.push('/mypage')} />
            </Box>
            <Drawer
                anchor={'left'}
                open={open}
                onClose={() => setOpen(false)}
                style={{
                    width: '600px'
                }}
                role='menubar'
            >
                {'drawer'}
            </Drawer>
        </Container>
    )
}

const Container = styled.div({
    display: 'flex',
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
    padding: 16,
    background: '#ffffff'
})