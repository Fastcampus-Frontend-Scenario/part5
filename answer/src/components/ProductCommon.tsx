/* eslint-disable @next/next/no-img-element */
import styled from "@emotion/styled"
import { Chip } from "@mui/material"
import { useRouter } from "next/router"
import { useMemo } from "react"
import { Product } from "types/types"

type PriceType = 'DISCOUNT' | 'PRICE'
type Props = {
    priceType: PriceType
    product: Product
}
export const ProductCommon: React.FC<Props> = ({ product, priceType }) => {
    const router = useRouter()
    const discounted = product.discount / product.price * 100
    const PriceComponent = useMemo(() => {
        switch(priceType) {
            case 'DISCOUNT': 
                return <Chip label={`${discounted}%`} color='error' />
            case 'PRICE': 
                return (
                    <PriceContainer>{`â‚©${product.price.toLocaleString()}`}</PriceContainer>
                )
        }
    }, [product, priceType, discounted])
    const handleClick = () => {
        router.push(`/product/${product.productId}`)
    }
    return (
        <Container onClick={handleClick}>
            <img src={product.imageSrc} alt={product.name} style={{ marginBottom: 4 }}/>
            <Title>{product.name}</Title>
            {PriceComponent}
        </Container>
    )
}

const Container = styled.div({
    display: 'flex',
    flexDirection: 'column',
    justifyContent: 'center',
    alignItems: 'center',
    minWidth: 140,
    minHeight: 180,
})

const Title = styled.div({
    fontSize: 13,
    color: '#1c1c1c',
    marginBottom: 8
})

const PriceContainer = styled.div({
    fontSize: 13,
    color: '#8B96A5'
})