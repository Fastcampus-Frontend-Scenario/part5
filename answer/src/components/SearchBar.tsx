import styled from "@emotion/styled"
import { Box } from "@mui/material"
import CloseIcon from '@mui/icons-material/Close';

import SearchOutlinedIcon from '@mui/icons-material/SearchOutlined';
import { useState } from "react";
import { useRouter } from "next/router";
export const SearchBar: React.FC = () => {
    const [value, setValue] = useState('')
    const router = useRouter()
    const handleSubmit = () => {
        const url = `/search?q=${value}`
        router.push(url)
    }
    return (
        <Container>
            <SearchBox>
                <SearchOutlinedIcon />
                <form onSubmit={handleSubmit} style={{flex: 1,}}>
                    <Input
                        placeholder="Search"
                        type='text'
                        value={value}
                        onChange={(e) => setValue(e.target.value)}
                    />
                </form> 
                {Boolean(value.length) && <CloseIcon onClick={() => setValue('')}/>}
            </SearchBox>
        </Container>
    )
}

const Container = styled.div({
    display: 'flex',
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
    padding: 16,
    background: '#ffffff'
})

const SearchBox = styled(Box)({
    display: 'flex',
    gap: '16px',
    flex: 1,
    alignItems: 'center',
    background: '#F7FAFC',
    borderRadius: 6,
    border: '1px solid #DEE2E7',
    padding: 8,
})

const Input = styled('input')({
    fontSize: 16,
    border: '0px',
    background: '#ffffff00',
    outline: 'none',
    flex: 1,
})