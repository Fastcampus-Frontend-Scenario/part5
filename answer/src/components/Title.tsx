import styled from "@emotion/styled";

export const Title = styled.p({
    fontSize: 18,
    fontWeight: 600,
    color: '#1c1c1c'

})