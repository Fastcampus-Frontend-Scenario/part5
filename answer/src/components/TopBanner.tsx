import axios from "axios"
import { useRouter } from "next/router"
import { BannerResponse } from "types/types"
import { useEffect, useState } from "react"

export const TopBanner = () => {
    const [banner, setBanner] = useState<BannerResponse | null>(null)
    const router = useRouter()

    const fetchBanner = async () => {
        const response = await axios.get<BannerResponse>('/api/ad-banner')
        const bannerData = response.data
        setBanner(bannerData)
    }
    useEffect(() => {
        void fetchBanner()
    }, [])
    const handleClick = () => {
        if (banner?.data?.url)
            router.push(banner?.data?.url)
    }
    if (!Boolean(banner) || banner?.result !== 'OK') return null

    return (
        // eslint-disable-next-line @next/next/no-img-element
        <div data-testid='top-banner-container'>
            <img src={banner.data?.imageSrc} onClick={handleClick} alt='ad-banner' data-testid='top-banner' />
        </div>
    )
}