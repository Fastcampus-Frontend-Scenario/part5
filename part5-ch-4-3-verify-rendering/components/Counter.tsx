import { useState } from "react"

type Props = {
    initialProps: number
}

const Counter: React.FC<Props> = ({ initialProps }: Props) => {
    const [count, setCount] = useState(initialProps)

    return (
        <div>
            <div role="display">{initialProps}</div>
            <br />
            <div role="count">{count}</div>
            <button name="+" onClick={() => setCount(count + 1)}>+</button>
            <button name="-" onClick={() => setCount(count - 1)}>-</button>
        </div>
    )
}

export default Counter