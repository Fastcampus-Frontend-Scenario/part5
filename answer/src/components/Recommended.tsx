import styled from "@emotion/styled"
import { Title } from "./Title"
import axios from "axios"
import { useRouter } from "next/router"
import { useState, useEffect } from "react"
import { RecommendedResponse } from "types/types"
import { ProductCommon } from "./ProductCommon"

export const Recommended: React.FC = () => {
    const [recommended, setRecommended] = useState<RecommendedResponse | null>(null)
    const router = useRouter()

    const fetcRecommended = async () => {
        const response = await axios.get<RecommendedResponse>('/api/recommended')
        const data = response.data
        setRecommended(data)
    }
    useEffect(() => {
        void fetcRecommended()
    }, [])

    if (!Boolean(recommended) || recommended?.result !== 'OK') return null
    return (
        <Container data-testid='recommended'>
            <Title>추천 제품들</Title>
            <ProductsContainer data-testid='recommended-product-list'>
                {recommended?.data.productList.map((product)=> {
                    return (
                        <ProductItem key={product.productId}>
                            <ProductCommon priceType='PRICE' product={product}/>
                        </ProductItem>
                    )
                })}
            </ProductsContainer>
        </Container>
    )
}

const Container = styled.div({
    display: 'flex',
    flexDirection: 'column',
    padding: 6,
    flex: 1,
})

const ProductsContainer = styled.div({
    display: 'grid',
    gridTemplateColumns: '1fr 1fr',
})

const ProductItem = styled.div({
    display: 'flex',
    borderRadius: 6,
    border: '1px solid #DEE2E7',
    margin: 4,
    background: '#ffffff'
})
