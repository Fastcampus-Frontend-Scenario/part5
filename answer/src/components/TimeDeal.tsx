import axios from "axios"
import { useRouter } from "next/router"
import { BannerResponse, TimeDealResponse } from "types/types"
import { useEffect, useMemo, useState } from "react"
import styled from "@emotion/styled"
import { ProductCommon } from "./ProductCommon"
import { Divider } from "@mui/material"
import { Title } from "./Title"

export const TimeDeal = () => {
    const [deal, setDeal] = useState<TimeDealResponse | null>(null)
    const router = useRouter()

    const fetcDeals = async () => {
        const response = await axios.get<TimeDealResponse>('/api/timedeal')
        const dealData = response.data
        setDeal(dealData)
    }
    useEffect(() => {
        void fetcDeals()
    }, [])

    if (!Boolean(deal) || deal?.result !== 'OK') return null
    const { items, dealEnded } = deal.data

    return (
        <Container data-testid='time-deal'>
            <TitleContainer>
                <Title>타임  딜</Title>
                <TimerComponent end={new Date(dealEnded)}/>
            </TitleContainer>
            <Divider style={{width: '100%'}} />
            <ProductsContainer data-testid='time-deal-product-container'>
                {items.map(item => {
                    return (<ProductCommon product={item} priceType='DISCOUNT' key={item.productId} />)
                })}
            </ProductsContainer>
        </Container>
    )
}

type TimerComponentProps = {
    end: Date
}
const TimerComponent: React.FC<TimerComponentProps> = ({ end }) => {
    const current = new Date()
    const [seconds, setSeconds] = useState<number>(Math.round((end.getTime() - current.getTime()) / 1000))
    const router = useRouter()

    useEffect(() => {
        const countdown = setInterval(() => {
            if (seconds > 0) {
                setSeconds(seconds - 1);
            }
            if (seconds === 0) {
                clearInterval(countdown)
            }
        }, 1000);
        return () => clearInterval(countdown);
    }, [seconds]);

    const remainSecounds = useMemo(() => Math.floor(seconds % 3600 % 60).toLocaleString('en-US', {
        minimumIntegerDigits: 2,
        useGrouping: false
    }), [seconds])
    const remainMinute = useMemo(() => Math.floor((seconds % 3600) / 60).toLocaleString('en-US', {
        minimumIntegerDigits: 2,
        useGrouping: false
    }), [seconds])
    const remainHour = useMemo(() => Math.floor(seconds / 3600).toLocaleString('en-US', {
        minimumIntegerDigits: 2,
        useGrouping: false
    }), [seconds])

    return (
        <TimerComponentContinaer>
            <TimeBlock><div style={{ fontSize: 16 }} data-testid='remain-hour'>{remainHour}</div><div style={{fontSize: 11}}>{'시간'}</div></TimeBlock>
            <TimeBlock><div style={{ fontSize: 16 }} data-testid='remain-min'>{remainMinute}</div><div style={{ fontSize: 11 }}>{'분'}</div></TimeBlock>
            <TimeBlock><div style={{ fontSize: 16 }} data-testid='remain-sec'>{remainSecounds}</div><div style={{ fontSize: 11 }}>{'초'}</div></TimeBlock>
        </TimerComponentContinaer>
    )
}

const Container = styled.div({
    display: 'flex',
    flexDirection: 'column',
    justifyContent: 'center',
    alignItems: 'center',
    padding: 10,
    background: '#ffffff'
})

const TitleContainer = styled.div({
    display: 'flex',
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
    width: '100%',
})


const ProductsContainer = styled.div({
    display: 'flex',
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'center',
    width: '100%',
    overflowY: 'auto'
})

const TimerComponentContinaer = styled.div({
    display: 'flex',
    flexDirection: 'row',
})

const TimeBlock = styled.div({
    display: 'flex',
    flexDirection: 'column',
    width: 45,
    height: 45,
    justifyContent: 'center',
    alignItems: 'center',
    background: '#EFF2F4',
    color: '#8B96A5',
    margin: 2,
})