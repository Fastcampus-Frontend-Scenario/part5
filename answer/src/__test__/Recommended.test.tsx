import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';
import { useRouter } from 'next/router';

import { screen } from '@testing-library/dom';
import { render, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { Recommended } from '../components/Recommended';

jest.mock('next/router', () => ({
    useRouter: jest.fn()
}))

describe('<Recommended>', () => {
    describe('Positive case', () => {
        let mock: MockAdapter
        beforeEach(() => {
            mock = new MockAdapter(axios)
            mock.resetHistory()
            mock.reset()
            const successObj = {
                'result': 'OK',
                'data': {
                    'productList': [
                        { 'productId': '1', 'name': 'Clothes', 'price': 30000, 'discount': 3000 },
                        { 'productId': '2', 'name': 'earphone', 'price': 100000, 'discount': 25000 },
                        { 'productId': '3', 'name': 'notebook', 'price': 1000000, 'discount': 250000 }
                    ]
                }
            }
            mock.onGet('/api/recommended').reply(200, successObj)
        })
        it('should render Recommended', async () => {
            render(<Recommended />)

            const recommended = await screen.findByTestId('recommended')
            expect(recommended).toBeInTheDocument()
        })

        it('should render recommended products correctly', async () => {
            render(<Recommended />)

            await screen.findByTestId('recommended')
            const container = screen.getByTestId('recommended-product-list')

            expect(container.childNodes.length).toBe(3)

            expect(screen.getByText('Clothes')).toBeInTheDocument()
            expect(screen.getByText('earphone')).toBeInTheDocument()
            expect(screen.getByText('notebook')).toBeInTheDocument()
        })

        it('should route to correct product page when item clicked', async () => {
            const user = userEvent.setup()
            const pushMock = jest.fn();

            (useRouter as jest.Mock).mockReturnValue({
                push: pushMock,
            })

            render(<Recommended />)

            await screen.findByTestId('recommended')
            await user.click(screen.getByText('Clothes'))
            expect(pushMock).toHaveBeenCalledWith(`/product/1`)
        })

        it('Render correctly', async () => {
            const { container } = render(<Recommended />)
            await screen.findByTestId('recommended')

            expect(container).toMatchSnapshot()
        })
    })
    describe('Negative case', () => {

        it('should not render top banner', async () => {
            const mock = new MockAdapter(axios)
            const errObj = {
                'result': 'ERROR',
            }
            mock.onGet('/api/recommended').reply(200, errObj)
            render(<Recommended />)
            await waitFor(() => expect(mock.history.get.length).toBe(1))

            expect(screen.queryByTestId('recommended')).not.toBeInTheDocument()
            mock.resetHistory()
        })
    })
    
})