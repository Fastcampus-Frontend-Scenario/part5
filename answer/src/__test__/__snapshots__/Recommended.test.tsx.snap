// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Recommended> Positive case Render correctly 1`] = `
<div>
  <div
    class="css-vdet7d"
    data-testid="recommended"
  >
    <p
      class="css-1fr51j6"
    >
      추천 제품들
    </p>
    <div
      class="css-ivrut9"
      data-testid="recommended-product-list"
    >
      <div
        class="css-lh2kmz"
      >
        <div
          class="css-mrwwdh"
        >
          <img
            alt="Clothes"
            style="margin-bottom: 4px;"
          />
          <div
            class="css-ocrbbw"
          >
            Clothes
          </div>
          <div
            class="css-k5ds6i"
          >
            ₩30,000
          </div>
        </div>
      </div>
      <div
        class="css-lh2kmz"
      >
        <div
          class="css-mrwwdh"
        >
          <img
            alt="earphone"
            style="margin-bottom: 4px;"
          />
          <div
            class="css-ocrbbw"
          >
            earphone
          </div>
          <div
            class="css-k5ds6i"
          >
            ₩100,000
          </div>
        </div>
      </div>
      <div
        class="css-lh2kmz"
      >
        <div
          class="css-mrwwdh"
        >
          <img
            alt="notebook"
            style="margin-bottom: 4px;"
          />
          <div
            class="css-ocrbbw"
          >
            notebook
          </div>
          <div
            class="css-k5ds6i"
          >
            ₩1,000,000
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
